src/java/org/apache/cassandra/index/sai/disk/io/IndexInputReader.java

```diff
@@ -0,0 +1,189 @@
+    public static IndexInputReader create(FileHandle handle)
+    {
+        RandomAccessReader reader = handle.createReader();
+        return new IndexInputReader(reader, () -> {});
+    }
+
+    @Override
+    public byte readByte() throws IOException
+    {
+        return input.readByte();
+    }
+
+    @Override
+    public void readBytes(byte[] bytes, int off, int len) throws IOException
+    {
+        try
+        {
+            input.readFully(bytes, off, len);
+        }
+        catch (CorruptBlockException ex)
+        {
+            throw new CorruptIndexException(input.getPath(), "Corrupted block", ex);
+        }
+    }
+
+    /**
+     * Using {@link RandomAccessReader#readShort()} directly is faster than {@link DataInput#readShort()} which calls
+     * {@link DataInput#readByte()} one by one
+     */
+    @Override
+    public short readShort() throws IOException
+    {
+        try
+        {
+            return input.readShort();
+        }
+        catch (CorruptBlockException ex)
+        {
+            throw new CorruptIndexException(input.getPath(), "Corrupted block", ex);
+        }
+    }
+
+    /**
+     * Using {@link RandomAccessReader#readInt()} directly is faster than {@link DataInput#readInt()} which
+     * calls {@link DataInput#readByte()} one by one
+     */
+    @Override
+    public int readInt() throws IOException
+    {
+        try
+        {
+            return input.readInt();
+        }
+        catch (CorruptBlockException ex)
+        {
+            throw new CorruptIndexException(input.getPath(), "Corrupted block", ex);
+        }
+    }
+
+    /**
+     * Using {@link RandomAccessReader#readLong()} directly is faster than {@link DataInput#readLong()} which
+     * calls {@link DataInput#readByte()} one by one
+     */
+    @Override
+    public long readLong() throws IOException
+    {
+        try
+        {
+            return input.readLong();
+        }
+        catch (CorruptBlockException ex)
+        {
+            throw new CorruptIndexException(input.getPath(), "Corrupted block", ex);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        try
+        {
+            input.close();
+        }
+        finally
+        {
+            doOnClose.run();
+        }
+    }
+
+    @Override
+    public long getFilePointer()
+    {
+        return input.getFilePointer();
+    }
+
+    @Override
+    public void seek(long position)
+    {
+        input.seek(position);
+    }
+
+    @Override
+    public long length()
+    {
+        return input.length();
+    }
+
+    @Override
+    public IndexInput slice(String sliceDescription, long offset, long length) throws CorruptIndexException
+    {
+        if (offset < 0 || length < 0 || offset + length > input.length())
+        {
+            throw new CorruptIndexException("Invalid slice! Offset: " + offset + ", Length: " + length + ", Input Length: " + input.length(), this);
```

Comment:
nit: Does failing one of these checks mean the index is corrupt, or just that we've passed silly arguments?

Aside: Should we throw the slice description in the error message?

Reply:
I've done a bit of investigation of this and it appears that `slice` is never called by us. It's a Lucene operation on `IndexInput` that we don't use. I've removed all this code and am now throwing an `UnsupportedOperationException` with a reasonable message.